example/service/s3/getObjectWithProgress/getObjectWithProgress.go

```diff
@@ -0,0 +1,158 @@
+// whilst also following the progress using the bytes written to the
+// writerAt object to track its progress
+type progressWriter struct {
+	writer  io.WriterAt
+	progressBar *pb.ProgressBar
+}
+
+func (pw *progressWriter) WriteAt(p []byte, off int64) (int, error) {
+	pw.progressBar.Add(len(p))
+	return pw.writer.WriteAt(p, off)
+}
+
+func byteCountDecimal(b int64) string {
+	const unit = 1000
+	if b < unit {
+		return fmt.Sprintf("%d B", b)
+	}
+	div, exp := int64(unit), 0
+	for n := b / unit; n >= unit; n /= unit {
+		div *= unit
+		exp++
+	}
+	return fmt.Sprintf("%.1f %cB", float64(b)/float64(div), "kMGTPE"[exp])
+}
+
+func getFileSize(svc *s3.S3, bucket string, prefix string) (filesize int64, error error) {
+	params := &s3.HeadObjectInput{
+		Bucket: aws.String(bucket),
+		Key: aws.String(prefix),
+	}
+
+	resp, err := svc.HeadObject(params)
+
+	if err != nil {
+		return 0, err
+	}
+
+	return *resp.ContentLength, nil
+}
+
+func validateS3URIString(s3Uri string) error {
+	hasMatch, err := regexp.MatchString("^s3://", s3Uri)
+	if err != nil {
+		return err
+	}
+	if !hasMatch {
+		return fmt.Errorf("%s not a valid S3 uri, Please enter a valid S3 uri. Ex: s3://mary/had/a/little/lamb", s3Uri)
+	}
+	return nil
+}
+
+func parseS3Uri(s3Uri string) (bucket string, prefix string) {
+	s3UriParts := strings.Split(s3Uri, "/")
+	prefix = strings.Join(s3UriParts[3:], "/")
+	bucket = s3UriParts[2]
+	return bucket, prefix
+}
+
+func parseFilename(keyString string) (filename string) {
+	ss := strings.Split(keyString, "/")
+	s := ss[len(ss)-1]
+	return s
+}
+
+func main() {
+	if len(os.Args) < 2 {
+		log.Println("USAGE ERROR: AWS_REGION=us-east-1 go run getObjWithProgress.go s3://example-s3-bucket/foo.zip")
+		return
+	}
+
+	s3Uri := os.Args[1]
+
+	err := validateS3URIString(s3Uri)
+
+	if err != nil {
+		panic(err)
+	}
+
+	bucket, key := parseS3Uri(s3Uri)
+
+	filename := parseFilename(key)
+
+	sess, err := session.NewSession()
+
+	s3Client := s3.New(sess)
+
+	downloader := s3manager.NewDownloader(sess)
+
+	size, err := getFileSize(s3Client, bucket, key)
+
+	if err != nil {
+		panic(err)
+	}
+
+	log.Printf("File size is: %s", byteCountDecimal(size))
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		panic(err)
+	}
+
+	temp, err := ioutil.TempFile(cwd, "getObjWithProgress-tmp-")
+
+	tempfileName := temp.Name()
+
+	if err != nil {
+		panic(err)
+	}
+
+	downloadBar := pb.New64(size).SetUnits(pb.U_BYTES)
+	downloadBar.Start()
+
+	writer := &progressWriter{writer: temp, progressBar: downloadBar}
+
+	params := &s3.GetObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+	}
+
+	if _, err := downloader.Download(writer, params); err != nil {
+		log.Printf("Download failed! Deleting tempfile: %s", tempfileName)
+		os.Remove(tempfileName)
+		panic(err)
+	}
+
+	downloadBar.FinishPrint(fmt.Sprintf("Downloaded %s", filename))
+
+	if err := os.Rename(temp.Name(), filename); err != nil {
+		panic(err)
+	}
+
+	if err := temp.Close(); err != nil {
+		panic(err)
```

Comment:
@jasdel I moved this lower and stuck some error logic around it, does that work for you?

Reply:
How the `Close` is called before the temp file is renamed?

Reply:
Done! 👍